src/note.ts

```diff
@@ -13,15 +13,254 @@
+
+export const GLYPH_PROPS_VALID_TYPES: Record<string, Record<string, string>> = {
+  n: { name: 'note' },
+  r: { name: 'rest' },
+  h: { name: 'harmonic' },
+  m: { name: 'muted' },
+  s: { name: 'slash' },
+};
+
+export interface Coordinates {
+  x: number;
+  y: number;
+}
+
+export interface Font {
+  glyphs: { x_min: number; x_max: number; ha: number; o: string[] }[];
+  cssFontWeight: string;
+  ascender: number;
+  underlinePosition: number;
+  cssFontStyle: string;
+  boundingBox: { yMin: number; xMin: number; yMax: number; xMax: number };
+  resolution: number;
+  descender: number;
+  familyName: string;
+  lineHeight: number;
+  underlineThickness: number;
+  /**
+   * This property is missing in vexflow_font.js, but present in gonville_original.js and gonville_all.js.
+   */
+  original_font_information?: {
+    postscript_name: string;
+    version_string: string;
+    vendor_url: string;
+    full_font_name: string;
+    font_family_name: string;
+    copyright: string;
+    description: string;
+    trademark: string;
+    designer: string;
+    designer_url: string;
+    unique_font_identifier: string;
+    license_url: string;
+    license_description: string;
+    manufacturer_name: string;
+    font_sub_family_name: string;
+  };
+}
+export interface GlyphProps {
+  code_head: string;
+  dot_shiftY: number;
+  position: string;
+  rest: boolean;
+  line_below: number;
+  line_above: number;
+  stem_up_extension: never;
+  stem_down_extension: never;
+  stem: never;
+  code: string;
+  code_flag_upstem: string;
+  code_flag_downstem: string;
+  flag: boolean;
+  width: number;
+  text: string;
+  tabnote_stem_down_extension: number;
+  tabnote_stem_up_extension: number;
+  beam_count: number;
+  duration_codes: Record<string, DurationCode>;
+  validTypes: Record<string, NameValue>;
+  shift_y: number;
+
+  getWidth(a?: number): number;
+
+  getMetrics(): GlyphMetrics;
+}
+
+export interface GlyphOptions {
+  fontStack: Font[];
+  category: string;
+}
+
+export interface GlyphMetrics {
+  width: number;
+  height: number;
+  x_min: number;
+  x_max: number;
+  x_shift: number;
+  y_shift: number;
+  scale: number;
+  ha: number;
+  outline: number[];
+  font: Font;
+}
+
+export interface Metrics {
+  totalLeftPx?: number;
+  totalRightPx?: number;
+  /** The total width of the note (including modifiers). */
+  width: number;
+  glyphWidth: number;
+  /** The width of the note head only. */
+  notePx: number;
+  /** Start `X` for left modifiers. */
+  modLeftPx: number;
+  /** Start `X` for right modifiers. */
+  modRightPx: number;
+  /** Extra space on left of note. */
+  leftDisplacedHeadPx: number;
+  glyphPx?: number;
+  /** Extra space on right of note. */
+  rightDisplacedHeadPx: number;
+}
+
+export interface NoteDuration {
+  duration: string;
+  dots: number;
+  type: string;
+}
+
+export interface DurationCode {
+  common: Type;
+  type: Record<string, Type>;
+}
+
+export interface Type extends KeyProps {
```

Comment:
Is there a better name than `Type` here?

Reply:
Being child of `KeyProps`, I changed to `TypeProps`, better than `Type` anyway.